src/v/pandaproxy/schema_registry/json.cc

```diff
@@ -602,11 +602,116 @@ json_type_list normalized_type(const json::Value& v) {
     return ret;
 }
 
-// helper to convert a boolean to a schema
+// helper to parse a json pointer with rapidjson. throws if there is an error
+// parsing it
+json::Pointer to_json_pointer(std::string_view sv) {
+    auto candidate = json::Pointer{sv.data(), sv.size()};
+    if (auto ec = candidate.GetParseErrorCode();
+        ec != rapidjson::kPointerParseErrorNone) {
+        throw as_exception(error_info{
+          error_code::schema_invalid,
+          fmt::format(
+            "invalid fragment '{}' error {} at {}",
+            sv,
+            ec,
+            candidate.GetParseErrorOffset())});
+    }
+
+    return candidate;
+}
+
+// helper to resolve a pointer in a json object. throws if the object can't be
+// retrieved
+const json::Value&
+resolve_pointer(const json::Pointer& p, const json::Value& root) {
+    auto unresolved_token = size_t{0};
+    auto* value = p.Get(root, &unresolved_token);
+    if (value == nullptr) {
+        throw as_exception(error_info{
+          error_code::schema_invalid,
+          fmt::format(
+            "object not found for pointer '{}' unresolved token at index "
+            "{}",
+            pjp{p},
+            unresolved_token)});
+    }
+
+    return *value;
+}
+
+// iteratively resolve a reference, following the $ref field until the end or
+// the max_allowed_depth is reached. throws if the max depth is reached or if
+// the reference can't be resolved
 json::Value::ConstObject
-get_schema(const schema_context&, const json::Value& v) {
+resolve_reference(schema_context& ctx, const json::Value& candidate) {
+    auto ref_it = candidate.FindMember("$ref");
+    if (ref_it == candidate.MemberEnd()) { // not a reference, no-op
+        return candidate.GetObject();
+    }
+
+    // we don't support merging of references with other fields
+    if (candidate.MemberCount() > 1) {
+        throw as_exception(error_info{
+          error_code::schema_invalid,
+          fmt::format(
+            "Unsupported merging of references with base object: '{}'",
+            pj{candidate})});
```

Comment:
This will give a subset of the doc, but I wonder if the `$ref` is more helpful?

Reply:
removed in the following commit set (kept here to support bisecting)